Realm/RLMResults.mm

```diff
@@ -563,6 +571,130 @@ - (RLMNotificationToken *)addNotificationBlock:(void (^)(RLMResults *, RLMCollec
+                                                                           subscriptionName:name
+                                                                                      query:_results.get_query()
+                                                                             updateExisting:true];
+            }];
+            [self completeOnQueue:queue completion:completion error:nil];
+            return;
+    }
+
+    [subscriptions updateOnQueue:queue block:^{
+        self.associatedSubscriptionId = [subscriptions addSubscriptionWithClassName:self.objectClassName
+                                                                   subscriptionName:name
+                                                                              query:_results.get_query()
+                                                                     updateExisting:true];
+    } onComplete:^(NSError *error) {
+        [self completeOnQueue:queue completion:completion error:error];
+    }];
+}
+
+// FIXME: Ultimately needs something cancellable from realm-core instead of sdk-level workaround.
+- (void)subscribeWithName:(NSString *_Nullable)name
+          waitForSyncMode:(RLMWaitForSyncMode)waitForSyncMode
+                  onQueue:(dispatch_queue_t _Nullable)queue
+                  timeout:(NSTimeInterval)timeout
+               completion:(RLMResultsCompletionBlock)completion {
+    // Create an internal completion that will only be called once
+    __block BOOL called = false;
+    void(^methodCompletion)(RLMResults* _Nullable, NSError* _Nullable) = ^(RLMResults* _Nullable results, NSError* _Nullable error) {
+        if (!called) {
+            if (error != nil) {
+                completion(nil, error);
+            } else {
+                completion(results, nil);
+            }
+            called = true;
+        }
+    };
+
+    // Setup timer
+    dispatch_time_t time =  dispatch_time(DISPATCH_TIME_NOW, (int64_t)(timeout * NSEC_PER_SEC));
+    // If the asynchronouns subscribe call below doesn't return after `time` seconds, the internal completion is called with an error.
+    dispatch_after(time, dispatch_get_main_queue(), ^{
+        NSString* errorMessage = [NSString stringWithFormat:@"Waiting for subscribed data timed out after %.01f seconds.", timeout];
+        NSError* error = [NSError errorWithDomain:RLMErrorDomain code:RLMErrorClientTimeout userInfo:@{NSLocalizedDescriptionKey: errorMessage}];
+        methodCompletion(nil, error);
+    });
+    // call asynchronous subscribe
+    [self subscribeWithName:name waitForSyncMode:waitForSyncMode onQueue:queue completion:methodCompletion];
+}
+
+- (void)unsubscribe {
+    RLMSyncSubscriptionSet *subscriptions = self.realm.subscriptions;
+
+    if (self.associatedSubscriptionId) {
+        [subscriptions update:^{
+            [subscriptions removeSubscriptionWithId:self.associatedSubscriptionId];
+        }];
+    } else {
```

Comment:
So if you have two subs with the same query. One named and one unnamed. And remove the named, the unnamed will stay in the set.
But yes in other cases you could have multiple results relying on the same subscription, if you remove one place it could affect the other results.